### DISCUSSION TITLE
Week 3: ML with Python

### DISCUSSION THREAD

**Discussion by BekahHW** [deepgram/community#87] [General help]
## Getting Started

The goal this week is to get to "Convolutional Neural Networks: Picking a Pretrained Model" lesson from [freeCodeCamp's ML with python](https://www.freecodecamp.org/learn/machine-learning-with-python/) and check-in here. 

## Learning Resources
- Introduction to ML for beginners: https://towardsdatascience.com/introduction-to-machine-learning-for-beginners-eed6024fdb08

### Neural Networks:
- https://www.3blue1brown.com/topics/neural-networks
- https://www.3blue1brown.com/lessons/gradient-descent

### Convolutional Neural Network
- [https://en.wikipedia.org/wiki/Convolutional_neural_network](https://en.wikipedia.org/wiki/Convolutional_neural_network)
- [https://hal.science/hal-02053205/document](https://hal.science/hal-02053205/document)
- [https://www.ibm.com/topics/convolutional-neural-networks](https://www.ibm.com/topics/convolutional-neural-networks)
- [https://towardsdatascience.com/a-comprehensive-guide-to-convolutional-neural-networks-the-eli5-way-3bd2b1164a53](https://towardsdatascience.com/a-comprehensive-guide-to-convolutional-neural-networks-the-eli5-way-3bd2b1164a53)

## Check-in
1. What have you learned about convolutional neural networks? 
2. How do you evaluate a CNN model?
3. What do you know today that you didn't when you started?
4. What are some key terms or ideas from these lessons?
5. What do you need to be clarified or what questions do you have?
6. What are you excited to learn more about?

**Comment by paceaux** [comment]
I've made it to "The Convolutional Layer"

1. what have I learned? the CNN is really the way to do deep learning. A CNN is just a very dense neural network with lots of layers
2. How do I evaluate a model? Not positive, but I think it involves creating a filter size, stride, and layers. 
3. What do I know now that I didn't at the beginning? The mechanics of how image detection detects ... stuff. Kinda. I think some of the mystery of how photoshop does its cropping and selecting is solved. 
4. key terms/ideas
    * filter
    * padding
    * stride
    * pooling
5. clarifications needed? I would've really liked, in his "convolutional layer" video, for him to explain how those numbers are actually calculated. Picking arbitrary numbers .. didn't help. 
6. Learn more about? I wonder how CNN works in non-image scenarios...

**Reply by paceaux** [reply]
Ok. Made it to "Picking a pretrained model"
1. What else have I learned? An easy way to train your image-detection model is to mess with the images. flip, scale, rotate, etc. Data augmentation is The Way 
2. How do I evaluate a model? I'm _still_ not sure I know, but I think I have a better idea:
     1. you need raw data and test data
     2. If it's an image, you gotta make sure they're all the same size
     3. If it's an image, you need to decide what you'll do about weights, and figure out your classes (labels)
     4. *if* you're using an existing model, you should freeze it
     5. Something-something add lots of layers
     6. Then you train it on your data
     7. then you save it
     8. Then you make a prediction with it
3. What do I know know that I didn't before:  At least for image detection, it's a lot about how many chunks of pixels you're willing to train at a time. Which makes it about layers. Models, like ogres and onions, have lots of layers. 
4. Clarifications? I want to know what this "trained model" actually looks like. Like, if we've created a "cats vs dogs" model ... I want to see how it's learned the difference. Where's _that_ ?
5. Learn more about? yeah I don't even know. Some of the mystery of object detection is gone for me, but this still feels a lot like dark magic.

**Comment by paceaux** [comment]
QQ: Is it possible to find this guy's actual notebooks that he's showing us?

**Reply by BekahHW** [reply]
Here are some of his colabs that I got from [this video](https://www.youtube.com/watch?v=tPYj3fFJGjk):

📕 Module 2: Introduction to TensorFlow - [https://colab.research.google.com/dri...](https://www.youtube.com/redirect?event=video_description&redir_token=QUFFLUhqbjNWVHA2S0ZfWk94bGRvUlRiOV9hWnRzVVZTUXxBQ3Jtc0ttOUlieklMRGdmVXh0N1FUVV9NUjBKWTJsWHJqSFF3MllTc3RZVVIyLWxhQ3ZnY2pfeHFGR3pQV0IyQXJPTE1SWGV0dzMyNTctWGw2a1pkTjhWQWRDdVlSejJGUlNiaDNuU3dPejhIbkMwVXV4bWZ6WQ&q=https%3A%2F%2Fcolab.research.google.com%2Fdrive%2F1F_EWVKa8rbMXi3_fG0w7AtcscFq7Hi7B%23forceEdit%3Dtrue%26sandboxMode%3Dtrue&v=tPYj3fFJGjk)
📗 Module 3: Core Learning Algorithms - [https://colab.research.google.com/dri...](https://www.youtube.com/redirect?event=video_description&redir_token=QUFFLUhqbW1ZaEI1bHlHckpxMU40NDhyZWkxVks0WHVVUXxBQ3Jtc0trdkVKcGJXdXJJWl9RV3huTFp0SGI4bGVnTE1qaFZ1alItSk9jT3NWREZZV3NfVDRGZms5Rm5RZkxDZXpHNmw2bW43NjZZUUlDczFtU1dhUVZ4dnBRUnUxbkpUelVaOEo0U2pLVzRRYXlNLWNUVGpuUQ&q=https%3A%2F%2Fcolab.research.google.com%2Fdrive%2F15Cyy2H7nT40sGR7TBN5wBvgTd57mVKay%23forceEdit%3Dtrue%26sandboxMode%3Dtrue&v=tPYj3fFJGjk)
📘 Module 4: Neural Networks with TensorFlow - [https://colab.research.google.com/dri...](https://www.youtube.com/redirect?event=video_description&redir_token=QUFFLUhqbUFlMl9XRnd0aHRTaU85SG9wNHAzN0lIX2JRZ3xBQ3Jtc0tsOEVWaWVWVGozclJZOU1TYzlGV0ZUbUg4XzZqemhxY2ZicEx6NTBvYjFlMWpaT01ZU0M4b3lUaFptTDU5amcwSk9zRnYtQWdHcFNNOHJQRjMzakpIWWpYWEhESHQ5U0FEcDVtMnBqdWxzZlJBQkJ2cw&q=https%3A%2F%2Fcolab.research.google.com%2Fdrive%2F1m2cg3D1x3j5vrFc-Cu0gMvc48gWyCOuG%23forceEdit%3Dtrue%26sandboxMode%3Dtrue&v=tPYj3fFJGjk)
📙 Module 5: Deep Computer Vision - [https://colab.research.google.com/dri...](https://www.youtube.com/redirect?event=video_description&redir_token=QUFFLUhqay1Rb25XdVh0X3JPc1Z4YUxFQ0p1cEJra2hsUXxBQ3Jtc0tsWmtkS1lrdy1uR0VmVS01WEkyVmhDY0dNLU02YWItcC1RVEtBMDZXY09jeGdxVGhzRkUwVHNZRm5FMDNGQmlBMDZYc3ZYOHNJZlZydFZKWjQ1QU1Fc25Fc0piX0pjZWtPTHdfTFhGa25WYklIYU83TQ&q=https%3A%2F%2Fcolab.research.google.com%2Fdrive%2F1ZZXnCjFEOkp_KdNcNabd14yok0BAIuwS%23forceEdit%3Dtrue%26sandboxMode%3Dtrue&v=tPYj3fFJGjk)
📔 Module 6: Natural Language Processing with RNNs -  [https://colab.research.google.com/dri...](https://www.youtube.com/redirect?event=video_description&redir_token=QUFFLUhqbWl5U09QSC1kb3d3U3Q3bGNUa3gydF8tUjQxUXxBQ3Jtc0trN2NkYjBSWkJIeW9lQ2xCazRnc1lQVktWYkZxWFlZbld3OEdOWmZVT2pscmR6WHhMbXdiWkdHN29KWTBnMXhHajJnbG04QW1paUc1ekhpQ3ctSEZHVHB1ZW1WTkdyVDhranNFd0RyaWJ5ajFjSVQtUQ&q=https%3A%2F%2Fcolab.research.google.com%2Fdrive%2F1ysEKrw_LE2jMndo1snrZUh5w87LQsCxk%23forceEdit%3Dtrue%26sandboxMode%3Dtrue&v=tPYj3fFJGjk)
📒 Module 7: Reinforcement Learning -  [https://colab.research.google.com/dri...](https://www.youtube.com/redirect?event=video_description&redir_token=QUFFLUhqbkx5NkJReUJhc0VNekNZelZzYzl0aWplSzl3QXxBQ3Jtc0trQ3A1UzRmRVhKc0gyRE5XeHFhcGlHY3gxWlk2T3pHYndZSTBfZ2J6YjVPc0JkN0FxNzJEazNXSGphd08xMEhQMTE5VkpCUnFkTVBnV19tZnN2Y0pTbWZIUjZWeEE5c2JDc2trdXdQVEdfV1F5aXQ3dw&q=https%3A%2F%2Fcolab.research.google.com%2Fdrive%2F1IlrlS3bB8t1Gd5Pogol4MIwUxlAjhWOQ%23forceEdit%3Dtrue%26sandboxMode%3Dtrue&v=tPYj3fFJGjk)

**Comment by BekahHW** [comment]
1. What have you learned about convolutional neural networks? They're most often used for image classification and object detection. There are three dimensions: height, width, color channels (rgb). The dense network looks at things globally. CNN find patterns anywhere in the image.
2. How do you evaluate a CNN model? Looking at how the model has been trained helps to understand the use case. The performance of the tests data set.
3. What do you know today that you didn't when you started?
4. What are some key terms or ideas from these lessons? CNN,  augmentation, sequential, keras,
5. What do you need to be clarified or what questions do you have? More on the architecture. 
6. What are you excited to learn more about? I'd like to explore more practical applications and the different open source models that are out there.
